Fix type inference involving wildcards

adriaanm opened the pull request:
Use == instead of =:= because we're looking for occurs, not equality.

When we were using =:= to detect occurrences of type params in bounds,
we'd accidentally unify too much (e.g. on wildcards / other type vars).

Discovered while looking into why this doesn't type check:
`java.util.stream.Stream.of(1,2,3).map(_.toString)`

I noticed that there was cross talk between the type params of java.util.Function during type inference.

This partially regressed in scala/scala#6492, but was actually wrong since the one with `*** empty log message ***`

Fix scala/bug#11558
Relates to scala/scala-dev#184

adriaanm committed:
Fix type infer regression introduced in scala/scala#6492

Use == instead of =:= because we're looking for occurs, not equality.

When we were using =:= to detect occurrences of type params in bounds,
we'd accidentally unify too much (e.g. on wildcards / other type vars).

Discovered while looking into why this doesn't type check:
`java.util.stream.Stream.of(1,2,3).map(_.toString)`

(scala/bug 11558)

I noticed that there was cross talk between the type
params of java.util.Function during type inference.

retronym commented:
Confirming that the problematic example from scala/scala-dev#184 progresses here.

Before:

```
scala> "abc".chars.boxed.map(x => x.toString).collect(java.util.stream.Collectors.toList())
                               ^
       error: type mismatch;
        found   : Integer => String
        required: java.util.function.Function[_ >: Integer, _]
```

After:

```
scala> "abc".chars.boxed.map(x => x.toString).collect(java.util.stream.Collectors.toList())
res4: java.util.List[String] = [97, 98, 99]
```

retronym commented:
Standalone test case that progresses:

```
trait F1[A, B] { def apply(a: A): B }; trait S[T] { def map[R](f: F1[_ >: T, _ <: R]): S[R] }; (f: S[Int]) => f.map(_.toString)
```

retronym commented on src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala:
```diff
@@ -240,26 +241,20 @@ private[internal] trait TypeConstraints {
-                tvar.addLoBound(tvarOther)
+          // Try to derive more constraints for `tvar` (and `tparam`) from its symmetric occurrences in the bounds of other tparams.
+          // `tparam` is the lower bound of `tvarOther`. Flip that, and add `tvarOther` as an upper bound for `tvar`.
+          // Use =:=, so that we equate eta-expanded type constructors (polytypes) and the equivalent no-arg typeref.
```
Is this comment out of date? I don't see an `=:=` below anymore.

adriaanm replied:
Yes, I’ll drop it

retronym commented:
Pulling a few more threads in https://github.com/retronym/scala/tree/review/8129

adriaanm committed:
Cleanups based on review feedback

Didn't take the typeSymbol-comparing route because
type constructor vars are too complicated to reason about
when moving constraints around -- just preserve old behavior.

adriaanm commented:
@retronym I incorporated a few strands of your review branch. LGTY?

SethTisue commented:
regression in overload resolution: https://github.com/scala/bug/issues/11755
